swap/src/api.rs

```diff
@@ -0,0 +1,311 @@
+use crate::protocol::Database;
+use crate::seed::Seed;
+use crate::{bitcoin, cli, monero};
+use anyhow::{Context as AnyContext, Result};
+use std::fmt;
+use std::net::SocketAddr;
+use std::path::PathBuf;
+use std::sync::{Arc, Once};
+use tokio::sync::{broadcast, Mutex};
+use url::Url;
+
+static START: Once = Once::new();
+
+#[derive(Clone, PartialEq, Debug)]
+pub struct Config {
+    tor_socks5_port: Option<u16>,
+    namespace: XmrBtcNamespace,
+    server_address: Option<SocketAddr>,
+    env_config: EnvConfig,
+    seed: Option<Seed>,
+    debug: bool,
+    json: bool,
+    data_dir: PathBuf,
+    pub is_testnet: bool,
+}
+
+// workaround for warning over monero_rpc_process which we must own but not read
+#[allow(dead_code)]
+pub struct Context {
+    pub db: Arc<dyn Database + Send + Sync>,
+    bitcoin_wallet: Option<Arc<bitcoin::Wallet>>,
+    monero_wallet: Option<Arc<monero::Wallet>>,
+    monero_rpc_process: Option<monero::WalletRpcProcess>,
+    running_swap: Arc<Mutex<bool>>,
+    pub config: Config,
+    pub shutdown: Arc<broadcast::Sender<()>>,
+}
+
+impl Context {
+    pub async fn build(
+        bitcoin: Option<Bitcoin>,
+        monero: Option<Monero>,
+        tor: Option<Tor>,
+        data: Option<PathBuf>,
+        is_testnet: bool,
+        debug: bool,
+        json: bool,
+        server_address: Option<SocketAddr>,
+        shutdown: broadcast::Sender<()>,
+    ) -> Result<Context> {
+        let data_dir = data::data_dir_from(data, is_testnet)?;
+        let env_config = env_config_from(is_testnet);
+
+        let seed = Seed::from_file_or_generate(data_dir.as_path())
+            .context("Failed to read seed in file")?;
+
+        let bitcoin_wallet = {
+            if let Some(bitcoin) = bitcoin {
+                let (bitcoin_electrum_rpc_url, bitcoin_target_block) =
+                    bitcoin.apply_defaults(is_testnet)?;
+                Some(Arc::new(
+                    init_bitcoin_wallet(
+                        bitcoin_electrum_rpc_url,
+                        &seed,
+                        data_dir.clone(),
+                        env_config,
+                        bitcoin_target_block,
+                    )
+                    .await?,
+                ))
+            } else {
+                None
+            }
+        };
+
+        let (monero_wallet, monero_rpc_process) = {
+            if let Some(monero) = monero {
+                let monero_daemon_address = monero.apply_defaults(is_testnet);
+                let (wlt, prc) =
+                    init_monero_wallet(data_dir.clone(), monero_daemon_address, env_config).await?;
+                (Some(Arc::new(wlt)), Some(prc))
+            } else {
+                (None, None)
+            }
+        };
+
+        let tor_socks5_port = tor.map(|tor| tor.tor_socks5_port);
+
+        START.call_once(|| {
+            let _ = cli::tracing::init(debug, json, data_dir.join("logs"), None);
+        });
+
+        let init = Context {
```

Comment:
Maybe rename this to `context`? I think you called this Init before but renamed the struct to Context. Up to you to decide